### ISSUE TITLE
Unable to detect error when extended attribute value is null！

### ISSUE THREAD

**Issue by SHECA-Alvin** [zmap/zlint#862]
### Description

We have a case in bugzilla where several error certificates contain empty cpsurls. We used zlint to check and no errors were reported. Please confirm whether zlint supports this error check. The zlint version used is v3.6.2
bugzilla address: https://bugzilla.mozilla.org/show_bug.cgi?id=1902947 

### appendix

The relevant error sample book is as follows:
-----BEGIN CERTIFICATE-----
MIIHfDCCBmSgAwIBAgIQdijhjRc6iD2nu+XtGDLX5DANBgkqhkiG9w0BAQsFADBA
MQswCQYDVQQGEwJDTjERMA8GA1UECgwIVW5pVHJ1c3QxHjAcBgNVBAMMFVNIRUNB
IEVWIFNlcnZlciBDQSBHMzAeFw0yNDAzMTgwNTM0MThaFw0yNTAzMTgxNTU5NTla
MIIBcDEdMBsGA1UEDxMUUHJpdmF0ZSBPcmdhbml6YXRpb24xEzARBgsrBgEEAYI3
PAIBAxMCQ04xGjAYBgsrBgEEAYI3PAIBAgwJ5LiK5rW35biCMRowGAYLKwYBBAGC
NzwCAQEMCeS4iua1t+W4gjEbMBkGA1UEBRMSOTEzMTAwMDA2MzEyOTEyODlYMQsw
CQYDVQQGEwJDTjESMBAGA1UECAwJ5LiK5rW35biCMRIwEAYDVQQHDAnkuIrmtbfl
uIIxDzANBgNVBBETBjIwMDA4MDFDMEEGA1UECQw65LiK5rW35biC6Jm55Y+j5Yy6
5Zub5bed5YyX6LevMTcxN+WPtzE4MDjjgIExODA544CBMTgxMOWupDE2MDQGA1UE
Cgwt5LiK5rW35biC5pWw5a2X6K+B5Lmm6K6k6K+B5Lit5b+D5pyJ6ZmQ5YWs5Y+4
MSIwIAYDVQQDExlyc2FldmczLnJldm9rZWQuc2hlY2EuY29tMIIBIjANBgkqhkiG
9w0BAQEFAAOCAQ8AMIIBCgKCAQEAo5512vvetdN4Yp0F5fuYit5yzmwwR9yHQNpt
Uu849hRMPfDcFtLTfXPwFHVTvFxPRiGmGIDoFrLnJ1A35Grt98B+avSk3tJFb6jz
tP/d2VboRpUgkicaU5Xqpg36rL0M59NGhqKkNu2wsc9+1B7dSqd3EQR2ufD0usPA
t7OJ1DXiJR+k8VpHBwVgm0y7shQ3JNOqjdT0jyz4bAcClJFQLWptvVPJbC1I7AZY
7//OBTuGAMD18Vi3fcH/stGitL/i0nMB4WQhpdi4XFIbWafviL1PPQsKi3RYpU72
n6a7YXCumK6x9odqt9pmGGwhUc96KfeZiTLWveI7OCEkJZonbwIDAQABo4IDPjCC
AzowHwYDVR0jBBgwFoAUVOly+3hmn+XL8zuPmEZVU3OcC4QwHQYDVR0OBBYEFBzW
siu+4iX0YZa4hQQzXBDENlaxMA4GA1UdDwEB/wQEAwIFoDAdBgNVHSUEFjAUBggr
BgEFBQcDAQYIKwYBBQUHAwIwXwYDVR0gBFgwVjA7BgkqgRyG7zoBAQMwLjAsBggr
BgEFBQcCARYgaHR0cHM6Ly93d3cuc2hlY2EuY29tL3JlcG9zaXRvcnkwFwYFZ4EM
AQEwDjAMBggrBgEFBQcCARYAMCQGA1UdEQQdMBuCGXJzYWV2ZzMucmV2b2tlZC5z
aGVjYS5jb20wDAYDVR0TAQH/BAIwADA4BgNVHR8EMTAvMC2gK6AphidodHRwOi8v
Y3JsLmdsb2JhbC5zaGVjYS5jb20vZXZzY2FnMy5jcmwwdwYIKwYBBQUHAQEEazBp
MDAGCCsGAQUFBzABhiRodHRwOi8vb2NzcC5nbG9iYWwuc2hlY2EuY29tL2V2c2Nh
ZzMwNQYIKwYBBQUHMAKGKWh0dHA6Ly9jZXJ0cy5nbG9iYWwuc2hlY2EuY29tL2V2
c2NhZzMuY2VyMIIBfwYKKwYBBAHWeQIEAgSCAW8EggFrAWkAdgDM+w9qhXEJZf6V
m1PO6bJ8IumFXA2XjbapflTA/kwNsAAAAY5QDbXBAAAEAwBHMEUCIQDvhPCwmHSC
R9NTYozUltUrpCrVDhIxGPm1zIHEiI5jbwIgUaSGy3b6TJQjBJ4i4bDeJlktk5am
UXjsFM0s4QqtITEAdgBOdaMnXJoQwzhbbNTfP1LrHfDgjhuNacCx+mSxYpo53wAA
AY5QDbXbAAAEAwBHMEUCIQC6ZOWdT938QKn5Yno620rj1asoyEaJuFw4PATORIjA
MQIgWmyLdTuuk8mIL0mSvkcfGIV0RROZBJ1KUIYTa8a6KHcAdwDPEVbu1S58r/OH
W9lpLpvpGnFnSrAX7KwB0lt3zsw7CAAAAY5QDbljAAAEAwBIMEYCIQDFPZECINh3
XHc/p0YSA2uXpVimDazVnDafXkB2nI8JYAIhAI0Wl+DkwJveSNrhX9GUHa54AxUd
fi12ms/GutfBow5DMA0GCSqGSIb3DQEBCwUAA4IBAQAxHBMMywdG+y9gECt7Fr4n
CPZsHFMfKgt1zzrCZHdUU1VKGMxLPfuJN9eRqOmnD5HCqFRCZHRl2nBIeB7tEdN+
3P4cHtIJK14hHlamYOig/iqLMY2ItkiT5XSgVLpyhGaZNmvgXYoD61X0IP1ueMnR
w2ivLGwrHtwbRGb0BD6zQN4uBfMPwn3J4TGaCWtG8kulDXJUyKfO1HRJD72tIzPn
zE9CRRRpZca27y2V8rSa7K+F5eF4N5ugUybkwrgq7WI/jvTyFE2XNtU57dI16hQ7
qtDi/RcbzJWXzstfDmk5y+KUvfe9bMsKtcBtRU+R5nak3d5EmMEdTl+5tzfyP0Ul
-----END CERTIFICATE-----

**Comment by zakird** [comment]
Hi @SHECA-Alvin! This issue feels like a great opportunity where you could help contribute to ZLint. ZLint is a community project and as it stands, your request comes across as directly offloading the work needed onto other CAs and contributors rather than making a good faith effort to understand what happened. I would encourage you to investigate whether an existing lint should have caught this. If not, can you suggest a lint that would help catch this in the future?

**Comment by SHECA-Alvin** [comment]
> Hi @SHECA-Alvin! This issue feels like a great opportunity where you could help contribute to ZLint. ZLint is a community project and as it stands, your request comes across as directly offloading the work needed onto other CAs and contributors rather than making a good faith effort to understand what happened. I would encourage you to investigate whether an existing lint should have caught this. If not, can you suggest a lint that would help catch this in the future?

Thank you for your reply. I will listen to your advice. Please allow me to do some research on zlint's code first. I am not very good at projects developed in golang language.

**Comment by mhyder13** [comment]
Hey all, I've done some research on this issue. It looks to me like this was resolved by #828. The current zlint codebase marks this certificate with an appropriate error `e_invalid_cps_uri`. I get this output from my local build:
```
| LEVEL | # OCCURRENCES |                       DETAILS                       |
+-------+---------------+-----------------------------------------------------+
|  info |             0 |                                                  -  |
|  warn |             2 |                      w_subject_common_name_included |
|       |               |w_ext_subject_key_identifier_not_recommended_subscri |
| error |             1 |                                   e_invalid_cps_uri |
| fatal |             0 |                                                  -  |
```

I reviewed the related bugs and found no citations for the premise that this is not allowed or details on the restriction. Going a bit deeper, I can see why this was not included in the original implementation pass. Arguably, an empty CPS URI is not really compliant with RFC 5280 but the paragraph describing the contents doesn't use clear normative language, which would have triggered its inclusion. 5280 § 4.2.1.4 describes it like this:
>  The CPS Pointer qualifier contains a pointer to a Certification
   Practice Statement (CPS) published by the CA.  The pointer is in the
   form of a URI.  Processing requirements for this qualifier are a
   local matter.  No action is mandated by this specification regardless
   of the criticality value asserted for the extension.

Looking at the then-current BRs, they probably further constrain this in §7.1.2.3(a), though again in a way that is hard to classify:
> The	following	extensions	MAY	be	present:	
certificatePolicies:policyQualifiers:policyQualifierId	(Recommended)	
• id‐qt	1	[RFC	5280].	
certificatePolicies:policyQualifiers:qualifier:cPSuri	(Optional)		
• HTTP	URL	for	the	Subordinate	CA's	Certification	Practice	

Again, this might have arguably breached the BRs even back in v1.4.8. At least, I wouldn't want to be a CA trying to argue against the point on bugzilla. I think it was not included as a requirement to check because it uses both "MAY" and "Optional." I can see that it didn't make it into our original spreadsheet of requirements to implement.

Looking now to the current BR version v2.1.5, the requirement is now a bit more clear. `policyQualifiers` is included in a table in §7.1.2.7.9 with the table labeled "Each PolicyInformation MUST match the following profile:." `policyQualifiers` then has a statement in the contents column "If present, MUST contain only permitted policyQualifiers from the table below." The table below indicates `id-qt-cps` is permitted with the contents column stating:
>The HTTP or HTTPS URL for the Issuing CA’s Certificate Policies,
Certification Practice Statement, Relying Party Agreement, or other
pointer to online policy information provided by the Issuing CA.

While that isn't quite the "If the CPS URI policyQualifier is present in a certificate, it MUST contain an HTTP or HTTPS URL" [recorded](https://github.com/zmap/zlint/blob/8696d6cdc9d509d22e455b82fc5a794d67d65a36/v3/lints/cabf_br/lint_e_invalid_cps_uri.go#L34) in lint `e_invalid_cps_uri`, I agree that this should be treated as a MUST requirement. By my reading, §7.1.2.7.9 is the section that applies to the certificate provided above, but several other sections address other types of certificates with identical tables and this additional statement which makes the intent clear.
> If the policyQualifiers is permitted and present within a PolicyInformation field,
it MUST be formatted as follows

While adding that label would have made the requirement more clear for subscriber certs, I think the two chained MUSTs in §7.1.2.7.9 serve to enforce the same requirement.

Looking at the commit history, it looks like `e_invalid_cps_uri` missed the v3.6.2 release by about a week and was included in [v3.6.3](https://github.com/zmap/zlint/releases/tag/v3.6.3). I did not attempt to reproduce this issue using v2.6.2.

Going back to 5280, From some brief searching, it appears to me that an empty string cannot be a valid URI. There's maybe some scope here to add a less-restrictive RFC-based lint from that language, but for public CAs involved in the web PKI, `e_invalid_cps_uri` should cover this structure for their purposes. I've also found a few lints that seem to reference old BR requirements that seem to have been rescinded, but I think that's best addressed in a fresh issue.

In conclusion, I believe this issue did exist, and that the standards support this lint, but this has been resolved with the addition of `e_invalid_cps_uri` in the v3.6.3 release. I'd like to recommend that this issue can likely be closed as Completed.
